## EDA (Exploratory Data Analysis)
In this section our goal is to derive some key insights from the data that we will be continuing to clean, transform and manipulate for our data analysis. We would like to understand what causes churn and what factors are contributing to higher churn rates if any. Some questions we would like to answer are:
1. Do customers who pay more lead to higher churn rates? (Numeric Features vs Churn)
2. Do customers who have a lower tenure tend to have higher churn rates? (Numeric Features vs Churn)
3. What categories are linked to higer churn rates? (Catergorical Features vs Churn)
4. Are certain combanations especially risky? (Cross Feature Insights)

In [899]:
import pandas as pd
import matplotlib

data = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")

We want to begin with locating the relevant data that we will be working with. 

In [900]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


When we run the code "data.info()" we recieve an output of some basic information from our data set. Taking a look at this we can say that some important insights are Tenure, Monthly Charges, Total Charges, Churn, Contract, Payment Method. Lets go ahead and create a data frame with this information. 

In [901]:
edaMetrics = data[["tenure","MonthlyCharges","TotalCharges","Churn","Contract","PaymentMethod", "InternetService"]]
edaMetrics.head()

tenure  MonthlyCharges TotalCharges Churn        Contract  \
0       1           29.85        29.85    No  Month-to-month   
1      34           56.95       1889.5    No        One year   
2       2           53.85       108.15   Yes  Month-to-month   
3      45           42.30      1840.75    No        One year   
4       2           70.70       151.65   Yes  Month-to-month   

               PaymentMethod InternetService  
0           Electronic check             DSL  
1               Mailed check             DSL  
2               Mailed check             DSL  
3  Bank transfer (automatic)             DSL  
4           Electronic check     Fiber optic

We are able to see that this information has now been organized into a data frame where we can visually see what we are doing. Lets begin by answering our first question. 
1. Do Customers who pay more tend to churn at a higher rate?

We can accomplish this by using a groupby statement and using our churn column as our index and "MonthlyCharges" column as our values to calculate churn by amount charged per month. We can then see the average amount paid based off whether a customer churned or stayed. 

In [902]:
"$" + edaMetrics.groupby("Churn")["MonthlyCharges"].mean().round(1).astype(str)

Churn
No     $61.3
Yes    $74.4
Name: MonthlyCharges, dtype: object

We are able to see that the average amount paid by customers who stayed was $64 and those who churned was $74 a month showing that these customers paid more on average. It is too risky to say that this was the sole cause of customers churning however it can be a cause so we should take this into account when conducting further churn analysis. 

We need to now examine our next question and revisit this calculation in further plotting and data analysis for churn. It is important to look at multiple angles to interpret the entire data set and give accurate reports. 

The next question we would like to answer is: 
1. Whether customers with lower tenure are more likely to churn? 
We can answer this by compairing lower tenure rates to higher tenure rates and seeing if lower tenure led to higher churn rates. 

In [903]:
edaMetrics["tenure"].describe()

count    7043.000000
mean       32.371149
std        24.559481
min         0.000000
25%         9.000000
50%        29.000000
75%        55.000000
max        72.000000
Name: tenure, dtype: float64

A great way to determing what type of data that we are working with is by using the ".describe()" method which gives us valuable insights toward the data we are working with. When we ran this command we are able to see various metrics that are being completed. One metric that immediatley speaks out to me is the max metric which depicts 72. This allows me to make the inference that tenure is most likely tracking months not days. However, it is imporant to note that while it is most likely months we should always attempt to ask the source as to what this data is measuring instead of guessing. The ".describe()" method is used to gather insight, not give you a for sure answer. 

In [904]:
edaMetrics.groupby("Churn")["tenure"].mean().round(0).astype(str) + " Months"

Churn
No     38.0 Months
Yes    18.0 Months
Name: tenure, dtype: object

We are able to see that the average tenure between customers that churned was 18 months or the equivalent of 1 year and half. This allows for us to understand that tenure may be playing a role in the churn rate for this customer database. We should continue along with our data analysis however, this is valuable information to consider as this may be a red flag indicator as to why customers are churning. 

In [905]:
edaMetrics["Contract"].describe()

count               7043
unique                 3
top       Month-to-month
freq                3875
Name: Contract, dtype: object

In [906]:
edaMetrics.head(15)

tenure  MonthlyCharges TotalCharges Churn        Contract  \
0        1           29.85        29.85    No  Month-to-month   
1       34           56.95       1889.5    No        One year   
2        2           53.85       108.15   Yes  Month-to-month   
3       45           42.30      1840.75    No        One year   
4        2           70.70       151.65   Yes  Month-to-month   
5        8           99.65        820.5   Yes  Month-to-month   
6       22           89.10       1949.4    No  Month-to-month   
7       10           29.75        301.9    No  Month-to-month   
8       28          104.80      3046.05   Yes  Month-to-month   
9       62           56.15      3487.95    No        One year   
10      13           49.95       587.45    No  Month-to-month   
11      16           18.95        326.8    No        Two year   
12      58          100.35       5681.1    No        One year   
13      49          103.70       5036.3   Yes  Month-to-month   
14      25          105.50      2686.05    No  Month-to-month   

                PaymentMethod InternetService  
0            Electronic check             DSL  
1                Mailed check             DSL  
2                Mailed check             DSL  
3   Bank transfer (automatic)             DSL  
4            Electronic check     Fiber optic  
5            Electronic check     Fiber optic  
6     Credit card (automatic)     Fiber optic  
7                Mailed check             DSL  
8            Electronic check     Fiber optic  
9   Bank transfer (automatic)             DSL  
10               Mailed check             DSL  
11    Credit card (automatic)              No  
12    Credit card (automatic)     Fiber optic  
13  Bank transfer (automatic)     Fiber optic  
14           Electronic check     Fiber optic

Using the ".describe()" method gives some insight as to what information is being contained in the Contract Column. We can see that there are 3 distinct values found within the Contract column. I use the ".head()" method to locate teh 3 distinct types of values that are held within the Contract Column. I can now continue with my data analysis. My question to answer is **Does contract type have an effect on churn rate?** This will show how categorical data affects churn rate.

We want to find out if contract length is affecting churn rate so the first thing we should do is compare each contract type and its churn response. 

In [907]:
pd.crosstab(edaMetrics["Contract"], edaMetrics["Churn"], normalize = "index").round(2).astype(str)+ " %"

Churn               No     Yes
Contract                      
Month-to-month  0.57 %  0.43 %
One year        0.89 %  0.11 %
Two year        0.97 %  0.03 %

I asked myself the question, **How can I compare two column categories where numerical columns are not being used but rather categorical data?** After researching how to best approach the problem I came accross the method "pd.crosstab" which essentially allows for proportions to take place which is great when working with categorical data. This function also allows for normalization. What this means is that we scale the numbers being on a 1% - 100% point scale rather then just comparing raw values which can be misleading when working with large datasets. Instead we allow for a proportion to take place. In this way we can see that out of all the month-to-month contract types 57% did not churn and 43% churned.

Now lets perform the same logic on the following categories:

PaymentMethod

Internet Service

In [908]:
pd.crosstab(edaMetrics["PaymentMethod"], edaMetrics["Churn"], normalize = "index").round(2).astype(str) + " %"

Churn                          No     Yes
PaymentMethod                            
Bank transfer (automatic)  0.83 %  0.17 %
Credit card (automatic)    0.85 %  0.15 %
Electronic check           0.55 %  0.45 %
Mailed check               0.81 %  0.19 %

I chose to compare Payment Methods and Churn to better understand how different types of payment can potentially be affecting the churn rate. Are bank transfers more effective at keeping customers or are Electronic Checks. The question is vital to help us answer the question of how to keep more customers. 

In [909]:
pd.crosstab(edaMetrics["InternetService"], edaMetrics["Churn"], normalize = "index").round(2).astype(str) + " %"

Churn                No     Yes
InternetService                
DSL              0.81 %  0.19 %
Fiber optic      0.58 %  0.42 %
No               0.93 %  0.07 %

I compare Internet Services vs Churn to see if the type of internet a customer holds affects their churn response. If a customer is paying for fiber optic internet services they must already have a high monthly payment. So this could be reflectory of their commitment to another monthly payment. 

This data is very insightful, and will allow for to make more accurate inferences as to what is causing customers to churn. As well as allow us to better understand how we can prevent it. Next we will tackle some visual interpretations to better analyze our data. 

In [910]:
avgMonthly = edaMetrics["MonthlyCharges"].mean().round(1)
hiMonthly = edaMetrics["MonthlyCharges"] > avgMonthly
loMonthly = edaMetrics["MonthlyCharges"] < avgMonthly


edaMetrics[hiMonthly]["Churn"].value_counts(normalize = True).round(2).astype(str)+ " %"

Churn
No     0.65 %
Yes    0.35 %
Name: proportion, dtype: object

In [911]:
edaMetrics[loMonthly]["Churn"].value_counts(normalize = True).round(2).astype(str)+ " %"

Churn
No     0.84 %
Yes    0.16 %
Name: proportion, dtype: object

What we can infer from these two compairisons is that customers with a higher monthly charge are 19% more likely to churn!